### PR TITLE
Prepare garbage collection uncommitted API

### LINKED ISSUE #4561: Create PrepareUncommittedForGC swagger API
Create the swagger API for PrepareUncommittedForGC

This is part of: https://github.com/treeverse/lakeFS/issues/4475

### PR DESCRIPTION
**Pull request by nopcoder**
Close #4561

### CODE REVIEW

File: api/swagger.yml [hunk 1]
```diff
@@ -981,6 +981,26 @@ components:
...
+      properties:
+        continuation_token:
+          type: string
+
+    PrepareGCUncommittedResponse:
+      type: object
+      properties:
+        run_id:
+          type: string
+        location:
```
Comment (N-o-Z) [inline api/swagger.yml:995]: We might want to add the location of the copy object file list to this command as well.
Suggest to change the name to be more specific

File: pkg/api/controller.go [hunk 2]
```diff
@@ -94,6 +95,91 @@ type Controller struct {
...
+			Action:   permissions.PrepareGarbageCollectionUncommittedAction,
+			Resource: permissions.RepoArn(repository),
+		},
+	}) {
+		return
+	}
+	ctx := r.Context()
+	c.LogAction(ctx, "prepare_garbage_collection_uncommitted", r, repository, "", "")
+
+	// TODO(barak): move this code the a Decode func in catalog
```
Comment (N-o-Z) [inline pkg/api/controller.go:133]: Is this comment still relevant?

File: pkg/api/controller.go [hunk 3]
```diff
@@ -2328,7 +2414,7 @@ func (c *Controller) PrepareGarbageCollectionCommits(w http.ResponseWriter, r *h
 	}
 	ctx := r.Context()
 	c.LogAction(ctx, "prepare_garbage_collection_commits", r, repository, "", "")
-	gcRUnMetadata, err := c.Catalog.PrepareExpiredCommits(ctx, repository, swag.StringValue(body.PreviousRunId))
+	gcRUnMetadata, err := c.Catalog.PrepareExpiredCommits(ctx, repository, StringValue(body.PreviousRunId))
```
Comment (N-o-Z) [inline pkg/api/controller.go:2417]:
```suggestion
	gcRunMetadata, err := c.Catalog.PrepareExpiredCommits(ctx, repository, StringValue(body.PreviousRunId))
```

File: pkg/api/controller.go [hunk 4]
```diff
@@ -94,6 +95,91 @@ type Controller struct {
...
+			WithFields(logging.Fields{"repository": repository, "continuation_token": continuationToken}).
+			Error("PrepareGCUncommitted failed")
+		c.handleAPIError(ctx, w, err)
+		return
+	}
+
+	// continuation token
+	var nextContinuationToken *string
+	if uncommittedInfo.Mark != nil {
+		nextMark, err := encodeGCUncommittedMark(uncommittedInfo.Mark)
```
Comment (N-o-Z) [inline pkg/api/controller.go:160]: Should remove the encode method from mark

File: pkg/catalog/catalog.go [hunk 5]
```diff
@@ -1871,8 +1857,11 @@ func (c *Catalog) PrepareGCUncommitted(ctx context.Context, repositoryID string,
 		return nil, err
 	}
 	defer func() {
-		fd.Close()
-		os.Remove(fd.Name())
+		_ = fd.Close()
+		err := os.Remove(fd.Name())
+		if err != nil {
+			c.log.WithField("filename", fd.Name()).Warn("Failed to delete temporary gc uncommitted data file")
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:1863]:
```suggestion
		if err := os.Remove(fd.Name()); err != nil {
			c.log.WithField("filename", fd.Name()).Warn("Failed to delete temporary gc uncommitted data file")
```

File: pkg/api/controller_test.go [hunk 6]
```diff
@@ -3083,3 +3083,90 @@ func TestController_GetPhysicalAddress_(t *testing.T) {
...
+				ContinuationToken: continuationToken,
+			})
+			verifyResponseOK(t, resp, err)
+			if resp.JSON201.RunId == "" {
+				t.Errorf("PrepareGarbageCollectionUncommittedWithResponse empty RunID, value expected")
+			}
+			if resp.JSON201.Location == "" {
+				t.Errorf("PrepareGarbageCollectionUncommittedWithResponse empty Location, value expected")
+			}
+			token := api.StringValue(resp.JSON201.ContinuationToken)
```
Comment (N-o-Z) [inline pkg/api/controller_test.go:3134]: We should validate that token == "" for items = 3, it is a bug if we get a different result

File: pkg/api/controller_test.go [hunk 7]
```diff
@@ -3083,3 +3083,90 @@ func TestController_GetPhysicalAddress_(t *testing.T) {
...
+			if resp.JSON201.Location != "" {
+				t.Errorf("PrepareGarbageCollectionUncommittedWithResponse Location=%s, expected empty", resp.JSON201.Location)
+			}
+			token := api.StringValue(resp.JSON201.ContinuationToken)
+			if token == "" {
+				break
+			}
+		}
+	})
+}
```
Comment (N-o-Z) [inline pkg/api/controller_test.go:3172]: Maybe add a check the ensures we get at least 1 cont token?

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Looks good!
Some comments / suggestions

File: api/swagger.yml (hunk 1 above)
Comment (nopcoder) [reply api/swagger.yml:995]: gc_uncommitted_location

File: pkg/api/controller.go (hunk 2 above)
Comment (nopcoder) [reply pkg/api/controller.go:133]: fixing

File: pkg/api/controller.go (hunk 4 above)
Comment (nopcoder) [reply pkg/api/controller.go:160]: Don't think there is Encode anymore

File: pkg/api/controller_test.go (hunk 6 above)
Comment (nopcoder) [reply pkg/api/controller_test.go:3134]: Endless loop, that will cause timeout - but I'll add verification for both committed and uncommitted.

File: pkg/api/controller_test.go (hunk 7 above)
Comment (nopcoder) [reply pkg/api/controller_test.go:3162]: Will depend on the catalog test for this one as generating 20MB using the API during tests will just slow the system down.
We will need to have this parameter configurable and test the token.

**Review by N-o-Z — APPROVED** [review]
LGTM!

**Pull request merged**